examples/graph/gnn_citations.py

```diff
@@ -0,0 +1,728 @@
+
+Note that the `graph_info` passed to the constructor of the Keras model, and used as a *property*
+of the Keras model object, rather than input data for training or prediction.
+The model will accept a **batch** of `node_indices`, which are used to lookup the
+node features and neighbours from the `graph_info`.
+"""
+
+
+class GNNNodeClassifier(tf.keras.Model):
+    def __init__(
+        self,
+        graph_info,
+        num_classes,
+        hidden_units,
+        aggregation_type="sum",
+        combination_type="concat",
+        dropout_rate=0.2,
+        normalize=True,
+        *args,
+        **kwargs,
+    ):
+        super(GNNNodeClassifier, self).__init__(*args, **kwargs)
+
+        # Unpack graph_info to three elements: node_features, edges, and edge_weight.
+        node_features, edges, edge_weights = graph_info
+
+        self.node_features = node_features
+        self.edges = edges
+        self.edge_weights = edge_weights
+        # Set edge_weights to ones if not provided.
+        if self.edge_weights is None:
+            self.edge_weights = tf.ones(shape=edges.shape[1])
+        # Scale edge_weights to sum to 1.
+        self.edge_weights = self.edge_weights / tf.math.reduce_sum(self.edge_weights)
+
+        self.num_classes = num_classes
+        self.hidden_units = hidden_units
+        self.aggregation_type = aggregation_type
+        self.combination_type = combination_type
+        self.dropout_rate = dropout_rate
+        self.normalize = normalize
+
+    def build(self, input_shape):
+
+        self.preprocess = create_ffn(
+            self.hidden_units, self.dropout_rate, name="preprocess"
+        )
+
+        self.conv1 = GraphConvLayer(
+            self.hidden_units,
+            self.dropout_rate,
+            self.aggregation_type,
+            self.combination_type,
+            self.normalize,
+            name="graph_conv1",
+        )
+
+        self.conv2 = GraphConvLayer(
+            self.hidden_units,
+            self.dropout_rate,
+            self.aggregation_type,
+            self.combination_type,
+            self.normalize,
+            name="graph_conv2",
+        )
+
+        self.skip_connection = layers.Add(name="skip_connection")
+
+        self.postprocess = create_ffn(
+            self.hidden_units, self.dropout_rate, name="postprocess"
+        )
+
+        self.compute_logits = layers.Dense(units=self.num_classes, name="logits")
+
+    def call(self, input_node_indicies):
+        # Preprocess the node_features to produce node representations.
+        x = self.preprocess(self.node_features)
+        # Apply the first graph conv layer.
+        x1 = self.conv1((x, self.edges, self.edge_weights))
+        # Skip connection.
+        x = self.skip_connection([x1, x])
+        # Apply the second graph conv layer.
+        x2 = self.conv2((x, self.edges, self.edge_weights))
+        # Skip connection.
+        x = self.skip_connection([x2, x])
```

Comment:
Just `+` is more readable (for all `skip_connection` calls). This isn't a Functional model.

I guess alternatively you *could* make this model a Functional model. Up to you.

Reply:
The reason I used an `Add` layer is that it displays  nicely (with a layer name "skip connection") when calling `model.summary` :-D

Reply:
If you want to keep this a subclassed model, then we should definitely use `+` in `call` rather than an `Add` layer. `Add` only makes sense in the context of a Functional model (addition as a layer), otherwise in `call` you should just use raw TF ops since you're not constructing a graph topology.

Reply:
Done